src/coreclr/nativeaot/Runtime/wasm/PalRedhawkWasm.cpp

```diff
@@ -39,4 +39,143 @@ void PalGetMaximumStackBounds_SingleThreadedWasm(void** ppStackLowOut, void** pp
     *ppStackLowOut = pStackLow;
     *ppStackHighOut = pStackHigh;
 }
+
+#ifdef TARGET_WASI
+// No-op stubs that assume a single-threaded environment.
+int pthread_mutex_init(pthread_mutex_t *, const pthread_mutexattr_t *)
+{
+    return 0;
+}
+
+int pthread_mutexattr_init(pthread_mutexattr_t *)
+{
+    return 0;
+}
+
+int pthread_mutexattr_settype(pthread_mutexattr_t *, int)
+{
+    return 0;
+}
+
+int pthread_mutex_destroy(pthread_mutex_t *)
+{
+    return 0;
+}
+
+int pthread_mutexattr_destroy(pthread_mutexattr_t *)
+{
+    return 0;
+}
+
+int pthread_cond_init(pthread_cond_t *, const pthread_condattr_t *)
+{
+    return 0;
+}
+
+int pthread_condattr_init(pthread_condattr_t *)
+{
+    return 0;
+}
+
+int pthread_mutex_lock(pthread_mutex_t *)
+{
+    return 0;
+}
+
+int pthread_mutex_unlock(pthread_mutex_t *)
+{
+    return 0;
+}
+
+pthread_t pthread_self(void)
+{
+    return (pthread_t)0;
+}
+
+int pthread_equal(pthread_t, pthread_t)
+{
+    return 1; // only one thread
+}
+
+int pthread_attr_init(pthread_attr_t *)
+{
+    return 0;
+}
+
+int pthread_attr_destroy(pthread_attr_t *)
+{
+    return 0;
+}
+
+int pthread_condattr_destroy(pthread_condattr_t *)
+{
+    return 0;
+}
+
+int pthread_cond_broadcast(pthread_cond_t *)
+{
+    return 0;
+}
+
+int pthread_attr_setdetachstate(pthread_attr_t *, int)
+{
+    return 0;
+}
+
+int mprotect(void *, size_t, int)
+{
+    return 0;
+}
+
+namespace __cxxabiv1 {
+
+    namespace {
+
+        using Dtor = void(*)(void*);
+
+        struct DtorList {
+            Dtor dtor;
+            void* obj;
+            DtorList* next;
+        };
+
+        // The linked list of thread-local destructors to run
+        DtorList* dtors = nullptr;
+
+        void run_dtors(void*) {
+            while (auto head = dtors) {
+                dtors = head->next;
+                head->dtor(head->obj);
+                ::free(head);
+            }
+        }
+
+        struct DtorsManager {
+            DtorsManager() {
+            }
+
+            ~DtorsManager() {
+                run_dtors(nullptr);
+            }
+        };
+    } // namespace
+
+    extern "C" int __cxa_thread_atexit(Dtor dtor, void* obj, void*)
+    {
+        static DtorsManager manager;
+
+        auto head = static_cast<DtorList*>(::malloc(sizeof(DtorList)));
+        if (!head) {
+            return -1;
+        }
+
+        head->dtor = dtor;
+        head->obj = obj;
+        head->next = dtors;
+        dtors = head;
+
+        return 0;
+    }
+} // namespace
```

Comment:
```suggestion
using Dtor = void(*)(void*);

// Due to a bug in the toolchain, we have to provide an implementation of thread-local destruction.
// Since this is the single-threaded case, we simply delegate to the static destruction mechanism.
// Reference: https://github.com/llvm/llvm-project/blob/main/libcxxabi/src/cxa_thread_atexit.cpp.
//
extern "C" int __cxa_thread_atexit(Dtor dtor, void* obj, void*)
{
    struct DtorList
    {
        Dtor dtor;
        void* obj;
        DtorList* next;
    };

    struct DtorsManager
    {
        DtorList* m_dtors = nullptr;

        ~DtorsManager()
        {
            while (DtorList* head = m_dtors)
            {
                m_dtors = head->next;
                head->dtor(head->obj);
                free(head);
            }
        }
    };

    // The linked list of "thread-local" destructors to run.
    static DtorsManager s_dtorsManager;

    DtorList* head = static_cast<DtorList*>(malloc(sizeof(DtorList)));
    if (head == nullptr)
    {
        return -1;
    }

    head->dtor = dtor;
    head->obj = obj;
    head->next = s_dtorsManager.m_dtors;
    s_dtorsManager.m_dtors = head;

    return 0;
}
```
I don't think we need to copy the namespace glue; but should conform to the runtime's conventions.

Reply:
Thanks, I was just going to add the comment for the origin, but you beat me to it.